### ISSUE TITLE
Return type mismatch not detected when implementing abstract methods.

### ISSUE THREAD

**Issue by AGiorgetti** [Microsoft/TypeScript#14796]
**TypeScript Version:**  2.2.1 

**Code**

```ts
class ReturnValue {
}

abstract class Base {
    abstract test1(arg: string): Promise<string[]>;
    abstract test2(arg: string): Promise<ReturnValue[]>;
}

class c extends Base {
    // with primitive type a mismath is detected (try change the return type to Promise<string>)
    test1(arg: string): Promise<string[]> {
        return null;
    }
    // it does not detect the mismatch on the return type
    test2(arg: string): Promise<ReturnValue> {
        return null;
    }
}
```

**Expected behavior:**

The compilation should fail: It should detect the return type mismatch of the 'test2' abstract method implementation.

**Actual behavior:**

The compilation succeeds.

**Comment by mhegazy** [comment]
This is not about abstract methods. this is about your type (`ReturnValue`) having no members. 

```ts
class Empty {
}


var x: Promise<Empty[]>;
var y: Promise<Empty>

x = y; // OK
y = x; // OK
```

Please see https://github.com/Microsoft/TypeScript/wiki/FAQ#why-do-these-empty-classes-behave-strangely for more details.

**Comment by AGiorgetti** [comment]
Oh you are right I just realized that all the types that were giving me this 'error' were all generated interfaces with optional dependencies.
This is working as intended, I'm closing the issue, sorry about that.
